modules/plugin/coverage-multidim/netcdf/src/main/java/org/geotools/coverage/io/netcdf/crs/NetCDFProjection.java

```diff
@@ -60,6 +60,139 @@
+                            projection.getOgcParameters(netcdfParameters),
+                            buildEllipsoid(var, SI.METRE));
+                };
+            };
+
+            /** Extract the CoordinateReferenceSystem from the provided objects */
+            abstract CoordinateReferenceSystem parseCRS(Variable variable, Attribute attribute)
+                    throws FactoryException;
+        }
+
+        private CRSParser(Attribute attribute, CRSParsingType type) {
+            this.attribute = attribute;
+            this.type = type;
+        }
+
+        /**
+         * Parse the Coordinate Reference System
+         *
+         * @param variable
+         * @return
+         * @throws FactoryException
+         */
+        public CoordinateReferenceSystem parseCoordinateReferenceSystem(Variable variable)
+                throws FactoryException {
+            return type.parseCRS(variable, attribute);
+        }
+
+        /**
+         * Extract the {@link CoordinateReferenceSystem} from a NetCDF attribute (if present)
+         * containing a WKT String
+         *
+         * @param wktAttribute the NetCDF {@link Attribute} if any, containing WKT definition.
+         * @return
+         */
+        private static CoordinateReferenceSystem parseWKT(Attribute wktAttribute) {
+            CoordinateReferenceSystem crs = null;
+            if (wktAttribute != null) {
+                String wkt = wktAttribute.getStringValue();
+                try {
+                    crs = CRS.parseWKT(wkt);
+                } catch (FactoryException e) {
+                    if (LOGGER.isLoggable(Level.WARNING)) {
+                        LOGGER.warning("Unable to setup a CRS from the specified WKT: " + wkt);
+                    }
+                }
+            }
+            return crs;
+        }
+
+        /**
+         * Get a proper {@link CRSParser} for the input Variable.
+         *
+         * @param var
+         * @return
+         */
+        public static NetCDFProjection.CRSParser getCRSParser(Variable var) {
+            if (var != null) {
+                Attribute attr = null;
+                // check on crs related attributes which may contain a fully
+                // defined WKT or a grid mapping reference
+                if ((attr = var.findAttribute(NetCDFUtilities.CERP_ESRI_PE_STRING)) != null) {
+                    return new WKTCRSParser(attr);
+                } else if ((attr = var.findAttribute(NetCDFUtilities.SPATIAL_REF)) != null) {
+                    return new WKTCRSParser(attr);
+                } else if ((attr = var.findAttribute(NetCDFUtilities.GRID_MAPPING_NAME)) != null) {
+                    return new GridMappingCRSParser(attr);
+                }
+            }
+            return null;
+        }
+
+        /** A CRSParser implementation based on a WKT String */
+        static class WKTCRSParser extends CRSParser {
```

Comment:
This is a bit strange, I undrestand you cannot have a simple enum because of parameters, but at this point, what's the value of the enum? I'd rather inline the code to do the parsing in these two subclasses. Mind, not a binding comment, I just find this construct confusing.

Reply:
Well. At the very beginning I was putting 4 enums to deal with different cases. But I managed to collapse 3 in 1... so yeah... I will simplify that.
Thanks

Reply:
+1